flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/NonReusingBuildFirstHashJoinIterator.java

```diff
@@ -144,13 +148,20 @@ public final boolean callWithNextKey(FlatJoinFunction<V1, V2, O> matchFunction,
 					// only single pair matches
 					matchFunction.join(nextBuildSideRecord, probeRecord, collector);
 				}
-			}
-			else if(joinWithEmptyBuildSide) {
-				// build side is empty, join with null
-				final V2 probeRecord = this.hashJoin.getCurrentProbeRecord();
+			} else {
+				if (probeSideOuterJoin && probeRecord != null && nextBuildSideRecord == null) {
```

Comment:
Can you add a few comments in this `else` branch? 
Please do the same for the other adapted HashJoinIterators.